pymeasure/instruments/hp/hplegacyinstrument.py

```diff
@@ -0,0 +1,218 @@
+
+c_uint8 = ctypes.c_uint8
+c_uint16 = ctypes.c_uint16
+c_uint32 = ctypes.c_uint32
+
+
+class StatusBitsBase(ctypes.BigEndianStructure):
+    """
+    A bitfield structure containing the assignments for the status decoding
+    """
+    _pack_ = 1
+    _get_process_ = {}
+    _set_process_ = {}
+
+    # decoder functions
+    # decimal to BCD & BCD to decimal conversion copied from
+    # https://pymodbus.readthedocs.io/en/latest/source/example/bcd_payload.html
+    @staticmethod
+    def _convert_from_bcd(bcd):
+        """Converts a bcd value to a decimal value
+
+        :param value: The value to unpack from bcd
+        :returns: The number in decimal form
+        """
+        place, decimal = 1, 0
+        while bcd > 0:
+            nibble = bcd & 0xF
+            decimal += nibble * place
+            bcd >>= 4
+            place *= 10
+        return decimal
+
+    @staticmethod
+    def _convert_to_bcd(decimal):
+        """Converts a decimal value to a bcd value
+
+        :param value: The decimal value to to pack into bcd
+        :returns: The number in bcd form
+        """
+        place, bcd = 0, 0
+        while decimal > 0:
+            nibble = decimal % 10
+            bcd += nibble << place
+            decimal /= 10
+            place += 4
+        return bcd
+
+    def __str__(self):
+        """
+        Returns a pretty formatted string showing the status of the instrument
+
+        """
+        ret_str = ""
+        for field in self._fields_:
+            ret_str = ret_str + f"{field[0]}: {hex(getattr(self, field[0]))}\n"
+
+        return ret_str
+
+    def __getattribute__(self, name):
+        val = super().__getattribute__(name)
+        if name == "fields":
+            return val
+        if name in self.fields():
+            process = super().__getattribute__('_get_process_')
+            if name in process:
+                val = process[name](val)
+
+        return val
+
+    def fields(self):
+        return [desc[0] for desc in super().__getattribute__('_fields_')]
+
+
+class PackedBitsBase(ctypes.BigEndianStructure):
+    """
+    A bitfield structure containing the assignments for the status decoding
+    """
+    _pack_ = 1
+    _get_process_ = {}
+    _set_process_ = {}
+
+    def __float__(self):
+        """
+        Return a float value from the packed data of the HP3437A
+
+        """
+        # range decoding
+        # (cf table 3-2, page 3-5 of the manual, HPAK document 9018-05946)
+        # 1 indicates 0.1V range
+        if self.range == 1:
+            cur_range = 0.1
+        # 2 indicates 10V range
+        if self.range == 2:
+            cur_range = 10.0
+        # 3 indicates 1V range
+        if self.range == 3:
+            cur_range = 1.0
+
+        signbit = 1
+        if self.sign_bit == 0:
+            signbit = -1
+
+        return (
+            cur_range * signbit * (
+                self.MSD + float(self.SSD) / 10 + float(self.TSD) / 100 + float(self.LSD) / 1000
```

Comment:
```suggestion
                self.MSD + self.SSD / 10 + self.TSD / 100 + self.LSD / 1000
```
I think the conversion to float is not needed. Right ?

Reply:
yes that was not needed